### ISSUE TITLE
Metadata for alias in elasticsearch connector only uses the first mapping 

### ISSUE THREAD

**Issue by brianolsen87** [prestosql/presto#2748]
In elasticsearch after creating an alias with multiple indices then requesting the mapping for all those indices returns a mapping for each .
```
curl --request POST \
  --url http://localhost:9200/_aliases \
  --header 'content-type: application/json' \
  --data '{
    "actions" : [
        { "add" : { "index" : "my_index1", "alias" : "my_index" } },
        { "add" : { "index" : "my_index2", "alias" : "my_index" } },
        { "add" : { "index" : "my_index3", "alias" : "my_index" } }
    ]
}'
```
and then request for the mapping
curl --request GET \
  --url http://localhost:9200/my_index/_mappings
I get back a response containing all the different indices.
```
{
 "my_index3": {<mapping3>},
 "my_index1": {<mapping1>},
 "my_index2": {<mapping2>}
 ...
}
```
Currently the elasticsearch connector will choose the mapping of the first index returned, in this case my_index3 mapping, and use that as the schema of all the indices. If my_index1 or my_index2 contain fields that aren't in my_index3, then those fields will not be available.

We could add a function that merges these mappings together for alias groups but there are some considerations.

1) should we add a limit to the number of indices we would merge? We could just leave it up to the user to know having too many indices in a search leads to poor performance.
2) I think adding a cache such as Cache<String, MetaData> with the index/alias as the key that expires after a minute or so would be good. Could also make this a configuration.

**Comment by martint** [comment]
I think we can have presto merge *all* mappings, but that will have to be supplemented by a cache as you suggested.

As for the question of what to do when types don't match, we should check what ES does when you run a query like that. Does it, effectively, behave as if the field was a union type?

**Comment by sumannewton** [comment]
@martint @bitsondatadev I have locally addressed all the points mentioned here except the caching. 
In brief:
Added 3 configs at the catalog level:

- union-schema-indices-for-alias: Flag to merge fields of all indices under an alias. Defaults to false(older behavior).
- fail-on-alias-schema-mismatch: Flag to decide whether to fail or not when there is a type mismatch for the same field name. Defaults to true. Gets into effect if _union-schema-indices-for-alias_ is true.
- default-datatype-for-alias-schema-mismatch: The default data type to be used in case of type mismatch for the same field name. Defaults to _text_(which eventually gets converted to varchar). Gets into effect if _fail-on-alias-schema-mismatch_ is false.

Let me know if this looks good. Will raise a PR if this is acceptable.

**Comment by bitsondatadev** [comment]
@sumannewton Awesome!! Really looking forward to your PR. If would say to submit the PR and we can make further comments there. 

Off the top of my head I would like for us to verify what @martint said by checking how ES handles mismatches so we can mirror that functionality. 
> As for the question of what to do when types don't match, we should check what ES does when you run a query like that. Does it, effectively, behave as if the field was a union type?

I may also push to add caching...it's rather simple and I had implemented this in my last company, we just never got it merged in before I left. It is just adding an expiring cache in the ElasticsearchMetaData I believe.

**Comment by sumannewton** [comment]
Yeah, caching is better. Will do that and raise a PR in a day or two.
Thanks @bitsondatadev

**Comment by sajjoseph** [comment]
@sumannewton - are you ready with your PR? 

Thanks!

**Comment by MusKaya** [comment]
This bug is also present when I do a simple select from indices with a wildcard. Such as:
`select * from my_index_*` selects from all matching indices such as `my_index_1`, `my_index_2`, ... however it assumes that all of them have the same mapping.

**Comment by bitsondatadev** [comment]
Hey @MusKaya,

These internally do the same thing. We can only produce a single mapping based on the requirements of SQL so when addressing multiple indices, we need to pull in some percentage or a sample of all the possible and merge them. 

Pulling the first index mapping is a simple heuristic to get us started but sampling multiple is the correct way to go. I call them both enhancements because there will really be no perfect performing way to handle this issue. Just a trade of that will have a parameter exposed to the user to tweak. This OSS next on the list once I finish my current elasticsearch fix.

**Comment by MusKaya** [comment]
@bitsondatadev I think by adding a `metastore-refresh-interval` (as in [hive connector](https://trino.io/docs/current/connector/hive.html#metastore-configuration-properties)) to the cache you have described above will also almost fully solve the latency issue I described in #7687

**Comment by subbareddydagumati** [comment]
Anyone working on this?

**Comment by bitsondatadev** [comment]
> Anyone working on this?

@sumannewton has a working PR but there are still some areas to fix.

If you are interested in taking over I can consult with him.
